File: neetcode/easy/reverse-a-linked-list.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def __str__(self):
        result = []
        current = self
        while current:
            result.append(str(current.val))
            current = current.next
        return " -> ".join(result)

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        while head:
            saved_next = head.next
            head.next = prev
            prev = head
            head = saved_next

        return prev

# Create the linked list with values [0, 1, 2, 3]
node3 = ListNode(3)
node2 = ListNode(2, node3)
node1 = ListNode(1, node2)
head = ListNode(0, node1)
print(Solution().reverseList(head))

File: neetcode/easy/reverse-a-linked-list.js
/**
 * Definition for singly-linked list.
 */
class ListNode {
  constructor(val = 0, next = null) {
    this.val = val;
    this.next = next;
  }
}

class Solution {
  /**
   * @param {ListNode} head
   * @return {ListNode}
   */
  reverseList(head) {
    let cur = head;

    let newNext = null;
    while (cur) {
      let temp = cur.next;
      cur.next = newNext;
      newNext = cur;
      cur = temp;
    }

    return newNext;
  }
}

// Helper function to create linked list from array
function createList(arr) {
  if (arr.length === 0) return null;
  let head = new ListNode(arr[0]);
  let current = head;
  for (let i = 1; i < arr.length; i++) {
    current.next = new ListNode(arr[i]);
    current = current.next;
  }
  return head;
}

// Create list from [0,1,2,3]
let list = createList([0, 1, 2, 3]);
console.log(list);
new Solution().reverseList(list);
